compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs

```diff
@@ -1045,6 +1047,78 @@ pub fn prohibit_assoc_item_binding(
             None
         },
     });
+
+    // Emit a suggestion if possible
+    if let Some((def_id, segment, _)) = segment
+        && segment.args().parenthesized == hir::GenericArgsParentheses::No
+        && let hir::TypeBindingKind::Equality { term: binding_arg_ty } = binding.kind
+    {
+        // Suggests removal of the offending equality constraint
+        let suggest_removal = |e: &mut Diag<'_>| {
+            let mut suggestion_span = binding.span.with_hi(binding.span.hi() + BytePos(1)); // Include the comma or the angle bracket at the end
+            if segment.args().bindings.len() == 1 {
+                // If it is the only binding specified
+                // include the starting angle bracket as well
+                suggestion_span = suggestion_span.with_lo(suggestion_span.lo() - BytePos(1))
+            }
+            if let Ok(suggestion) = tcx.sess.source_map().span_to_snippet(suggestion_span) {
+                e.span_suggestion_verbose(
+                    suggestion_span,
+                    "try removing this type binding",
+                    suggestion,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        };
+
+        // Suggests replacing the quality constraint with
+        // normal type argument
+        let suggest_direct_use = |e: &mut Diag<'_>, sp: Span, is_ty: bool| {
+            if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(sp) {
+                let ty_or_const = if is_ty { "generic" } else { "const" };
+                e.span_suggestion_verbose(
+                    binding.span,
+                    format!("to use `{snippet}` as a {ty_or_const} argument specify it directly"),
+                    snippet,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        };
+
+        // Get a sense of what generic args the type expects
+        let generics = tcx.generics_of(def_id);
+        let GenericParamCount { mut types, consts, .. } = generics.own_counts();
```

Comment:
Instead of looking at `own_counts` I'm wondering if we should just iterate through `generics.params` and look for a param def where the name matches. Remember: The motivation for this suggestion is suggesting `PartialEq<Type>` when encountering `Trait<Rhs = Type>` (param name == `Rhs` == binding name).

Reply:
Yeah, that'll be nicer. I'll look into it.

Reply:
Following this suggestion, I've changed the logic now to check for a param with a matching name.